### PR TITLE
Add an optional torch backend: the same numbers, differentiable

### PR DESCRIPTION
**Pull request by estcarisimo**
> Part 16 of a 20-commit stack. Stacked on #37 — review that first; the diff here against its base is a single commit.

levy/backends/ gains a torch implementation of the interpolation and the tail
approximation. It is a reimplementation, not a wrapper -- the NumPy code writes
into arrays under boolean masks, which autograd cannot follow, so the same
arithmetic is expressed with torch.where, torch.stack and out-of-place indexing.

What it buys: a stable distribution can sit inside a larger torch model and be
fitted by gradient descent along with everything else, instead of only through
this package's own L-BFGS-B. Hand pdf/cdf/logpdf a tensor and you get a tensor
back with gradients attached; levy.set_backend('torch') and levy.using('torch')
select it explicitly.

Two claims, both measured rather than asserted:

* Parity. Contract is rtol=1e-6 over a grid spanning the interpolated region
  and both tails, six (alpha, beta, mu, sigma) settings, pdf/cdf/neglog. What
  is actually observed is max relative deviation of 2.6e-16 on macOS/x86 and
  1.4e-12 on the Linux CI runner -- float64 rounding and a libm difference,
  because both paths do the same arithmetic over the same float32 grid but
  torch and NumPy do not order the 64-term interpolation sum identically. A
  second test asserts 1e-10 -- two orders of magnitude over the worse of those
  measurements, four below the contract -- so that if the two ever genuinely
  diverge the loose 1e-6 does not hide it.

* Gradients. torch.autograd.gradcheck passes in float64 for all four
  parameters, for pdf, cdf and neglog, at three parameter settings. And it is
  useful, not just correct: 400 Adam steps on a 4000-point sample land on the
  same optimum scipy's L-BFGS-B finds, to within 5e-3 --

      gradient descent  [1.576, 0.355, 0.466, 1.233]
      L-BFGS-B          [1.576, 0.355, 0.466, 1.233]
      truth             [1.6,   0.3,   0.5,   1.2  ]

Two things are deliberately not differentiable, and the module docstring says
so: the integer grid cell used to look up the crossover limits, and the mask it
implies. Both are choices, not quantities -- the crossover is piecewise
constant in alpha -- so the gradient is right everywhere except exactly at a
cell boundary, where a step function has no derivative anyway.

Tensor parameters in a parametrization other than 0 are refused, with the
reason: the conversion formulas run in NumPy and would cut the gradient without
saying so. Float parameters in any parametrization are fine, because there is
nothing to cut.

If torch is absent, not one line of it is imported. tests/test_no_torch.py
proves that in a subprocess with torch blocked at the import system: the whole
API works, the whole 1.x surface works, importing levy.backends does not pull
torch in, and selecting it fails at set_backend -- where the mistake was made,
not three calls later -- with `pip install "pylevy[torch]"`.

fit() is deliberately left on scipy. The value here is a differentiable
objective you can drop into your own optimizer; rerouting fit through
torch.optim would change its numerics for everyone to no purpose.

levy/backends/_numpy.py is a pure re-export. The core was not reorganised to
accommodate a second backend, so adding torch cannot have moved a NumPy number
-- and the install-parity check confirms it: 99,312 values bit-identical
against the parent commit.

57 new tests (862 total). Goldens unchanged.

Co-Authored-By: Claude Opus 5 (1M context) <noreply@anthropic.com>

### CODE REVIEW

File: src/levy/api.py [hunk 1]
```diff
@@ -418,7 +524,11 @@ def pdf(
     >>> np.round(pdf(np.array([1.0, 2.0]), alpha=1.5, beta=0.0), 6)
     array([0.202038, 0.084539])
     """
-    p = _validated(alpha, beta, mu, sigma, par)
+    module = _dispatch(backend, par, x, alpha, beta, mu, sigma)
+    p = _validated(_plain(alpha), _plain(beta), _plain(mu), _plain(sigma), par)
+    if module.name != 'numpy':
+        return _from_backend(module.pdf(x, alpha, beta, mu, sigma))
+
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/api.py:630]: Confirmed and fixed (cc0160b). `_for_backend` now hands a non-NumPy backend the validated, converted value for every parameter that is not a tensor, and the caller's own object for one that is -- so gradients survive in parametrization 0, and a call in 1, M, A or B is evaluated as requested. `_dispatch` already refuses tensors outside parametrization 0, so the two cases cannot mix.

Pinned by `test_float_parameters_in_another_parametrization_are_converted`: pdf, cdf and logpdf in each of the four other parametrizations, against the NumPy path. All 12 cases failed before the fix.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -242,16 +242,21 @@ jobs:
...
       # half: pandas installed, so tests/test_pandas.py actually runs instead of
       # being skipped.
+      # CPU-only: the torch wheel index below avoids pulling ~2 GB of CUDA
+      # libraries onto a runner that has no GPU.
       - name: Install package with extras
         run: |
           python -m pip install --upgrade pip
-          python -m pip install -e ".[dev,pandas]"
+          python -m pip install torch --index-url https://download.pytorch.org/whl/cpu
+          python -m pip install -e ".[dev,pandas,torch]"
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply .github/workflows/ci.yml:272]: Taken (bcc2846), both halves. A step reads `[project].name` from pyproject.toml and asserts through importlib.metadata that the `pandas` and `torch` extras are declared -- that catches a dropped or misspelt extra regardless of what is pre-installed, which is the right check given that the CPU-wheel pre-install has to stay. And the not-skipped step now runs both optional suites once with `-rs` and fails on any SKIPPED line, so an import failure that made pytest skip would fail the job instead of passing it.

File: src/levy/api.py [hunk 3]
```diff
@@ -418,7 +556,11 @@ def pdf(
...
     """
-    p = _validated(alpha, beta, mu, sigma, par)
+    module = _dispatch(backend, par, x, alpha, beta, mu, sigma)
+    p = _validated(_plain(alpha), _plain(beta), _plain(mu), _plain(sigma), par)
+    if module.name != 'numpy':
+        return _from_backend(module.pdf(x, *_for_backend(p, alpha, beta, mu, sigma)))
+
     labels = labels_of(x)
     values = _levy(np.asarray(x, dtype='d') if labels else x,
                    p.alpha, p.beta, p.mu, p.sigma, cdf=False)
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/api.py:633]: Taken (bcc2846). `_dispatch` refuses pandas input on a non-NumPy backend with a ValueError that says why -- the labels cannot be put back on a tensor -- and what to do: `x.to_numpy()` for a tensor, or NumPy parameters to keep the labels. Before this it died inside `torch.as_tensor` with "could not determine the shape of object type 'Series'". The three docstrings now say pandas input is NumPy-only, and `test_pandas_input_on_the_torch_backend_is_refused_with_a_reason` covers both routes to the torch backend.

File: src/levy/backends/_torch.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+* the grid cell used to look up the tail crossover limits, which is an integer
+  index derived from ``alpha`` and ``beta``;
+* the mask separating the interpolated region from the power-law tail, which
+  follows from those limits.
+
+A gradient with respect to ``alpha`` therefore ignores the fact that moving
+``alpha`` far enough would move the crossover. Over any distance a gradient step
+cares about, that term is zero anyway -- the crossover is piecewise constant in
+``alpha`` -- so the gradient is right almost everywhere and wrong only exactly
+at a cell boundary, where a step function has no derivative to begin with.
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/backends/_torch.py]: Taken (bcc2846): the note now says the mask compares (x - mu)/sigma against the limits and is therefore a step function in x, mu and sigma as well, and the paragraph after it is worded for any argument rather than alpha alone.

File: src/levy/backends/_torch.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    -----
+    The floor at 1e-100 matches the NumPy path. It is applied with
+    :func:`torch.clamp`, whose gradient is zero for clamped entries -- the right
+    answer, since a density the table reports as non-positive carries no
+    information about which way to move.
+    """
+    torch = _torch()
+
+    return -torch.log(torch.clamp(
+        _evaluate(x, alpha, beta, mu, sigma, cdf=False), min=1e-100))
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/backends/_torch.py]: Confirmed and taken (62cd4dd): the floor is `max(1e-100, torch.finfo(dtype).tiny)`, so float64 keeps the NumPy floor exactly and float32 gets the smallest positive normal instead of a clamp at zero. `test_float32_neglog_stays_finite_where_the_density_underflows` evaluates x = 1e16, 1e20 and 1e30 in float32 -- all three returned inf before -- and checks float64 is unchanged against NumPy.

File: tests/test_torch.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    # order or vectorise it identically. Measured worst case is 2.6e-16 on
+    # macOS/x86 and 1.4e-12 on the Linux CI runner -- a libm difference, not a
+    # divergence. 1e-10 keeps two orders of magnitude of headroom over the
+    # worse of those while staying four below the contract.
+    expected = numpy_backend.pdf(XS, alpha, beta, mu, sigma)
+    got = torch_backend.pdf(torch.tensor(XS, dtype=DOUBLE),
+                            alpha, beta, mu, sigma).numpy()
+    finite = np.abs(expected) > 1e-300
+    relative = np.abs(got[finite] - expected[finite]) / np.abs(expected[finite])
+    assert relative.max() < 1e-10, f"max relative deviation {relative.max():.3e}"
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply tests/test_torch.py:90]: The description was stale; the code and its comment were right. Updated the PR body to state what is measured (2.6e-16 on macOS/x86, 1.4e-12 on the Linux runner -- a libm difference in how torch and NumPy order the 64-term sum) and that the tighter test asserts 1e-10, two orders of magnitude over the worse measurement and four below the contract. Nothing in the code changed for this.

File: src/levy/api.py [hunk 7]
```diff
@@ -366,6 +367,117 @@ def _narrow(value: Any) -> ScalarOrArray:
...
+    Returns
+    -------
+    float or ndarray
+        Declared, not checked. The same localised inaccuracy as
+        :func:`_labelled`, and for the same reason: torch is an optional extra
+        and must not become a type-checking dependency. Every call a type
+        checker will normally see uses the NumPy backend and really does return
+        float or ndarray.
+    """
+    return cast(ScalarOrArray, value)
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/api.py:490]: Declining, deliberately, and the reason is the one written in `_from_backend`'s docstring: torch is an optional extra, and tensor-typed overloads would make it a type-checking dependency of the whole package -- the lint job's mypy run does not have torch installed, and should not need to. The `py.typed` contract is the NumPy surface; the torch path is a runtime dispatch, and a caller who type-checks torch code casts at the call site. If there is demand for typed overloads they belong in a follow-up that also decides how the lint environment gets torch, not in the commit that introduces the backend.

File: src/levy/api.py [hunk 8]
```diff
@@ -570,7 +743,11 @@ def logpdf(
     >>> np.round(logpdf(np.array([1.0, 2.0]), alpha=1.5, beta=0.0), 6)
     array([-1.599299, -2.470541])
     """
-    p = _validated(alpha, beta, mu, sigma, par)
+    module = _dispatch(backend, par, x, alpha, beta, mu, sigma)
+    p = _validated(_plain(alpha), _plain(beta), _plain(mu), _plain(sigma), par)
+    if module.name != 'numpy':
+        return _from_backend(-module.neglog(x, *_for_backend(p, alpha, beta, mu, sigma)))
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/api.py:765]: Taken (b5fa620): the `logpdf` docstring now states the floor per dtype -- 1e-100 (about -230) for NumPy and float64 tensors, the dtype's smallest positive normal for lower precision, about 1.2e-38 or -87 in float32, where 1e-100 is not representable.

File: README.md [hunk 9]
```diff
@@ -78,6 +78,33 @@ api.fit(returns).to_series()
...
+
+sample = torch.tensor(observations)
+params = torch.tensor([1.4, 0.0, 0.0, 1.0], requires_grad=True)
+optimizer = torch.optim.Adam([params], lr=0.03)
+
+for _ in range(400):
+    optimizer.zero_grad()
+    loss = -api.logpdf(sample, alpha=params[0], beta=params[1],
+                       mu=params[2], sigma=params[3]).sum()
+    loss.backward()
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply README.md]: Taken (b5fa620): the README example clamps alpha, beta and sigma the way the optimisation test does, with a comment saying why, and I ran it as written on a 2000-point sample -- it converges.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -280,21 +283,24 @@ jobs:
...
+          python -m pytest tests/ -q -m "not build"
+          --ignore=tests/test_pandas.py --ignore=tests/test_torch.py
+
+      # -rs prints every skip. Each optional suite skips itself when its
+      # dependency is missing, so any SKIPPED line means an extra did not
+      # install -- which pre-installing torch above would otherwise hide.
+      - name: Optional-dependency suites are exercised, not skipped
         run: |
-          python -m pytest tests/test_pandas.py -q -rs -p no:cacheprovider | tee optional.log
+          python -m pytest tests/test_pandas.py tests/test_torch.py -q -rs -p no:cacheprovider | tee optional.log
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply .github/workflows/ci.yml:306]: Taken (691ec05), the same `set -o pipefail` as on the pandas step it extends.

File: src/levy/backends/_torch.py [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+    ``scipy.special.gamma(alpha)`` becomes ``exp(lgamma(alpha))``, which is
+    differentiable in ``alpha`` and, for the range this package covers, equal to
+    it to within rounding. The in-place masked writes of the NumPy version
+    become :func:`torch.where`, which is what autograd can follow.
+    """
+    torch = _torch()
+
+    positive = x > 0
+    magnitude = (torch.sin(math.pi * alpha / 2.0) * torch.exp(torch.lgamma(alpha))
+                 / math.pi * torch.abs(x) ** (-alpha - 1.0))
```
[the remark this answers is not in the document]
Comment (estcarisimo) [reply src/levy/backends/_torch.py]: Confirmed and taken (1a6ac3f). The tail coefficient's sine is computed in float64 and cast back to the tensor's dtype -- float32 pi's sine is -8.7e-8, float64's is +1.2e-16, and the sign is the whole difference -- with the reason in a comment. `test_float32_tail_at_alpha_2_is_not_negative` checks the float32 pdf is non-negative and the cdf at most one at alpha = 2 (before: -4.5e-13 and 1.0 by clamping), and that the gradient with respect to alpha is still finite through the two casts.
